pkg/quarterdeck/middleware/ratelimiter.go

```diff
@@ -0,0 +1,123 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/api/v1"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/config"
+	"golang.org/x/time/rate"
+)
+
+/*
+IPRateLimiter is an IP address based limiter that controls how frequently requests
+can be made from a single IP address.
+limit: represents the number of tokens that can be added to the token bucket per second
+burst: maximum number of tokens/requests in a "token bucket" and is initially full
+each request consumes tokens from the token bucket and if the bucket is empty
+when the request is made, the request is rejected
+*/
+type IPRateLimiter struct {
+	ips   map[string]*ipInfo //contains a map of IP address to information about its request pattern
+	mu    *sync.RWMutex
```

Comment:
Rather than make this a `*sync.RWMutex` I'm going to embed it into the struct as a value; this makes it so the mutex can't be copied, which is the source of some subtle bugs.